File: lab/javascript/textos/verificar_parentesis.js
class Pilha {
    constructor() {
        this.pilha = [];
    }

    isEmpty() {
        return this.pilha.length === 0;
    }

    push(x) {
        this.pilha.push(x);
    }

    pop() {
        if (this.isEmpty()) {
            return null;
        }
        return this.pilha.pop();
    }

    peek() {
        if (this.isEmpty()) {
            return null;
        }
        return this.pilha[this.pilha.length - 1];
    }
}


function verifica_parenteses(string) {
    const pilha = new Pilha();

    for (const char of string) {  // <-- importante: for...of
        if (char === '(' || char === '[' || char === '{') {
            pilha.push(char);
        } else if (char === ')' || char === ']' || char === '}') {
            if (pilha.isEmpty()) {
                return false; // fechou sem abrir
            }
            pilha.pop();
        }
    }

    return pilha.isEmpty(); // true se estiver balanceado
}


// Testes
const strings = [
    "(())",
    "(()",
    "())",
    "",
    "((()()))"
];

for (const s of strings) {  // <-- for...of
    console.log(`${s} -> ${verifica_parenteses(s) ? 'Válido' : 'Inválido'}`);
}

File: lab/python/textos/verificar_parentesis.py
class Pilha:
    def __init__(self):
        self.pilha = []

    def isEmpty(self):
        return not self.pilha  # True se estiver vazia

    def push(self, x):
        self.pilha.append(x)

    def pop(self):
        if self.isEmpty():
            return None
        return self.pilha.pop()

    def peek(self):
        if self.isEmpty():
            return None
        return self.pilha[-1]


def verifica_parenteses(string):
    pilha = Pilha()

    for char in string:
        if char == '(' or char == "{" or char == "[":
            pilha.push(char)
        elif char == ')' or char == "}" or char == "]":
            if pilha.isEmpty():  # Tentou fechar sem abrir
                return False
            pilha.pop()

    # No fim, se sobrou algo, está errado
    return pilha.isEmpty()


# Testes
strings = [
    "(())",       # balanceado
    "(()",        # falta fechar
    "())",        # fechou demais
    "",           # vazio (ok)
    "((()()))"    # balanceado
]

for s in strings:
    print(f"{s:10} -> {'Válido' if verifica_parenteses(s) else 'Inválido'}")
